README.md
# WARNING
This is an edit of prev libftest made by @jtoty to fit with the 2019 project libft. <br/>
(ORIGINAL REP : https://github.com/jtoty/Libftest) <br/>
This update was made by @lmartin <br/>
Some things may not work (especially bonus part with lst functions), and things aren't tested. <br/>
I will try to update this (@lmartin).

# Libftest
<img align="right" src="./srcs/img_readme.png" width="50%"/>
This repository contains script to test your libft project.
The script will do the following tests :

- Check if the author file exists
- Check content of the author file
- Check if libft.h exists
- Check norme of libft.h
- Check if the Makefile file exists
- Check rules in Makefile
- Check compilation's flag in Makefile
- Compil library
- Check norme errors
- Check forbidden functions
- Tests functions

At the end of the tests, a deepthought file will be created, inside which you
can find all the results and error/failure messages. You can also see your
results in tests directory.

The script will compil your files with your libft, so you must have your
Makefile up to date.

:warning:All the tests made are not the official tests:warning:

## Getting Started

### Installation

```bash
git clone https://github.com/jtoty/Libftest
```

### Configuration

If you run the script for the first time, it will automatically create
`my_config.sh` file and ask you to edit it with the path of your libft
project.
You can also choose the colors that will be used and the path where the
deepthought file will be created.

## Running script

Go to the directory where you cloned and run the script
```bash
bash grademe.sh
```
or simply run the script with the path
```bash
bash /path/where/you/cloned/grademe.sh
```

### Options available
| Option | Description |
| --- | --- |
| `-h`<br />`--help` | Display help and exit |
| `-d` | Allows to perform the tests even if the files are in directories |
| `-c` | Disable color |
| `-s` | Disable searching Makefile and author files |
| `-m` | When compiling library, test all the Makefile's rules (instead of doing only make re and checking if other rules exist). |
| `-l` | Disable compiling library |
| `-f` | Disable checking forbidden functions |
| `-n` | Disable norminette |
| `-u` | Disable checking updates at launch |
| `-p1` | Disable part 1 tests |
| `-p2` | Disable part 2 tests |
| `-b`| Disable bonus part tests |
| `-a` | Disable additional part tests |
| `-op1`| Do only part 1 tests |
| `-op2`| Do only part 2 tests |
| `-ob` | Do only bonus part tests |
| `-oa` | Do only additional part tests |
| `ft_function` | Test only this function |

Add successively all options you want, in the order you want.
For example :
```bash
bash grademe.sh ft_atoi -f ft_strlen -n
```

### Supported functions
All the supported functions are listed [on this page](https://github.com/jtoty/Libftest/blob/master/supported_functions.md).
## Contribution

Any suggestions or bugs reporting ?
Contact jtoty@student.42.fr

### Contributors
- jmichaud : https://github.com/MrJe
- tlernoul


changelog
~~~~~~~~~~~~~~~~~~~~~~~~~~2019/01/17~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Error in ft_striteri
- Error in ft_strmapi

~~~~~~~~~~~~~~~~~~~~~~~~~~2018/06/29~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strtok

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/11/16~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strlcat

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/11/13~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Error in ft_strlcpy
- Error in ft_lstdelone

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/09/25~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strcasestr

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/04/14~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strncasecmp
- One test for ft_strrchr

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/04/02~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Legend at the end of the script

#Fixed
- Tests for fd functions

#Changed
- Option -q becomes -m and allows to test all the Makefile's
rules instead of doing only make re

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/28~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Abort detection
- Bus error detection
- Timeout detection

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/25~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Display format of -h/--help option

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/02/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Test 1 of ft_strnequ

~~~~~~~~~~~~~~~~~~~~~~~~~~2017/01/23~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Option -d : allows to perform the tests even if the files
are in directories

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/22~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strsub

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/20~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_itoa
- One test for ft_itoa_base

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/19~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_itoa_base

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/16~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strchr
- One test for ft_strrchr

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Option -op1 : do only part 1 tests
- Option -op2 : do only part 2 tests
- Option -ob : do only bonus part tests
- Option -oa : do only additional part tests
- Tests for ft_lstadd
- Tests for ft_lstdelone
- Tests for ft_lstdel

#Changed
- Tests of ft_strchr
- Tests of ft_strrchr
- Tests of ft_strmap
- Tests of ft_strmapi
- Tests of ft_striter
- Tests of ft_striteri

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/04~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_memmove

#Changed
- Test 1 of ft_strcat
- Test 3 of ft_strcat
- Test 1 of ft_strncat
- Test 2 of ft_strncat
- Test 5 of ft_strncat
- Test 1 of ft_strlcat
- Test 3 of ft_strlcat
- Test 4 of ft_strlcat
- Test 5 of ft_strlcat
- Test 6 of ft_strlcat
- Test 8 of ft_strlcat

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/12/03~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strncpy

#Changed
- Image for README.md

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/30~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- New character when display tests

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/30~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One test for ft_strlcat
- One test for ft_bzero

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/29~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Display error in deepthought file

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/21~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Forbidden functions detection. nm command is used now for
better detection. Option -f to disable it
- List of forbidden functions in part 1
- Improve rapidity when checking norme

#Fixed
- Bug when checking forbidden functions

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/21~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Bug when checking forbidden functions

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/20~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- One additional function : ft_strcasecmp

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/19~~~~~~~~~~~~~~~~~~~~~~~~~~

#Changed
- Test 2 of ft_strnequ (previous test was wrong)
- Test 3 of ft_strnequ (previous test was wrong)
- Test 5 of ft_strnequ (previous test was wrong)
- Test 6 of ft_strnequ (previous test was wrong)

#Added
- Option -h / --help : display help and exit
- One additional function : ft_strndup

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Remove
- Tests for ft_lstadd, ft_lstdel & ft_lstdelone due to errors

#Added
- One test for ft_strlcat
- Display name of each part in deepthought file

#Changed
- Test 13 of ft_atoi
- Test 2 of ft_atoi
- Test 18 of ft_atoi

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Tests for ft_lstadd
- One test for ft_strnstr
- One test for ft_strlcat
- Tests for ft_lstdel
- Tests for ft_lstdelone
- One test for ft_strlcpy
- ft_*.o are now removed at the end of the tests
- One test for ft_strnew

#Changed
- Test 5 of ft_strlcat
- Test 1 of ft_strnew

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/08~~~~~~~~~~~~~~~~~~~~~~~~~~

#Fixed
- Tests for ft_strnequ

#Added
- Tests for ft_lstnew
- Tests for ft_lstiter
- Tests for ft_lstmap

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/07~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Additional test for ft_strlcat
- Additional test for ft_strcmp
- Additional test for ft_strtrim
- Additional test for ft_strncmp
- Additional test for ft_strsplit

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/05~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Additional test for ft_memmove

#Changed
- One test of ft_memcmp
_ One test of ft_strnew

#Fixed
- Checking content of author file

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/04~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Additional test for ft_isascii
- Checking if libft.h exists
- Checking for norme erros in libft.h
- Tests for ft_split, ft_memchr.

#Changed
- Separate mandatory functions in two parts : Part1 and Part2
- Image for README.md

~~~~~~~~~~~~~~~~~~~~~~~~~~2016/11/03~~~~~~~~~~~~~~~~~~~~~~~~~~

#Added
- Checking if Makefile file exists
- Checking Makefile's rules
- Checking compilation's flags in Makefile
- Checking if author file exists
- Checking content of author file
- Tests for all functions of the mandatory part
- Checking for forbidden functions
- Checking for norme errors
- Checking for updates at launch

grademe.sh
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    grademe.sh                                         :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jtoty <jtoty@student.42.fr>                +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2017/01/23 18:26:01 by jtoty             #+#    #+#              #
#    Updated: 2019/10/13 19:29:55 by lmartin          ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

#!/bin/bash

#PATH_TEST=$(dirname $(readlink -f $0))
PATH_TEST="$(cd "$(dirname "$0")" && pwd -P)"

#exec 6>&2 2>/dev/null
exec 2> /dev/null

source ${PATH_TEST}/srcs/variables.sh

for arg in ${@}
do
	case "${arg}" in
		"--help")			man ${PATH_TEST}/srcs/help.1
							exit ;;
		"-h")				man ${PATH_TEST}/srcs/help.1
							exit ;;
		"-d")				DIRECTORY=1 ;;
		"-s")				OPT_NO_SEARCH=1 ;;
		"-m")				OPT_FULL_MAKEFILE=1 ;;
		"-l")				OPT_NO_LIBRARY=1 ;;
		"-c")				OPT_NO_COLOR=1 ;;
		"-f")				OPT_NO_FORBIDDEN=1 ;;
		"-n")				OPT_NO_NORMINETTE=1 ;;
		"-u")				OPT_NO_UPDATE=1 ;;
		"-op1")				OPT_NO_PART2=1
							OPT_NO_BONUS=1
							OPT_NO_ADDITIONAL=1
							CHECK_IN_PART1=0
							CHECK_IN_PART2=0
							CHECK_IN_BONUS=0
							CHECK_IN_ADDITIONAL=0 ;;
		"-op2")				OPT_NO_PART1=1
							OPT_NO_BONUS=1
							OPT_NO_ADDITIONAL=1
							CHECK_IN_PART2=0
							CHECK_IN_PART1=0
							CHECK_IN_BONUS=0
							CHECK_IN_ADDITIONAL=0 ;;
		"-ob")				OPT_NO_PART1=1
							OPT_NO_PART2=1
							OPT_NO_ADDITIONAL=1
							CHECK_IN_BONUS=0
							CHECK_IN_PART2=0
							CHECK_IN_PART1=0
							CHECK_IN_ADDITIONAL=0 ;;
		"-oa")				OPT_NO_PART1=1
							OPT_NO_PART2=1
							OPT_NO_BONUS=1
							CHECK_IN_ADDITIONAL=0
							CHECK_IN_BONUS=0
							CHECK_IN_PART2=0
							CHECK_IN_PART1=0 ;;
		"-p1")				OPT_NO_PART1=1
							CHECK_IN_PART1=0 ;;
		"-p2")				OPT_NO_PART2=1
							CHECK_IN_PART2=0 ;;
		"-b")				OPT_NO_BONUS=1
							CHECK_IN_BONUS=0 ;;
		"-a")				OPT_NO_ADDITIONAL=1
							CHECK_IN_ADDITIONAL=0 ;;
		*ft_*)	for part in ${tab_all_part[*]}
				do
					check_part=$(echo CHECK_IN_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
					if [ ${!check_part} -eq 1 ]
					then
						#opt_part=$(echo OPT_NO_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
						p=0
						activate_part=$(echo ACTIVATE_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
						tab_part=$(echo ${part}[*])
						nb_func=$(echo ${!tab_part} | wc -w)
						while (( p < ${nb_func} ))
						do
							func_name=$(echo ${part}[$p])
							if [ "$(echo ${arg} | grep -ow $(echo ${!func_name} | cut -d . -f 1))" != "" ]
							then
								(( ${part}_activation[$p]=1 ))
								#(( ${opt_part}=1 ))
								#####################
								OPT_NO_PART1=1
								OPT_NO_PART2=1
								OPT_NO_BONUS=1
								OPT_NO_ADDITIONAL=1
								#############################
								(( ${activate_part}=1 ))
								break
							fi
							(( p += 1 ))
						done
					fi
				done ;;
	esac
done

source ${PATH_TEST}/srcs/colors.sh
source ${PATH_TEST}/srcs/check_cheat.sh
source ${PATH_TEST}/srcs/check_compilation.sh
source ${PATH_TEST}/srcs/check_file.sh
source ${PATH_TEST}/srcs/check_norme.sh
source ${PATH_TEST}/srcs/compil_lib.sh
source ${PATH_TEST}/srcs/diff_test.sh
source ${PATH_TEST}/srcs/test_function.sh
source ${PATH_TEST}/srcs/check_update.sh

cd ${PATH_TEST}

if [ ${OPT_NO_UPDATE} -eq 0 ]
then
	func_check_update
fi

if [ ! -e ${PATH_TEST}/my_config.sh ]
then
	printf "${BOLD}my_config.sh${DEFAULT} file is not found.\n"
	printf "Creating file...\n"
	if [ -e ${PATH_TEST}/srcs/config_template.sh ]
	then
		cp ${PATH_TEST}/srcs/config_template.sh ${PATH_TEST}/my_config.sh
		printf "File created with success in ${BOLD}${PURPLE}${PATH_TEST}\n${DEFAULT}"
		printf "${RED}${UNDERLINE}Edit my_config.sh file${DEFAULT} with the path of your libft project and launch script.\n"
	else
		printf "Can't create my_config.sh file, try to update or clone again the repository and retry.\n"
		exit
	fi
	exit
fi

source ${PATH_TEST}/my_config.sh

if [ ${DIRECTORY} -eq 1 ]
then
	if [ -d ${PATH_TEST}/dirlibft ]
	then
		rm -rf ${PATH_TEST}/dirlibft
	fi
	printf "Copying files...\nPlease wait a moment.\n"
	mkdir ${PATH_TEST}/dirlibft
	cp -r ${PATH_LIBFT}/* ${PATH_TEST}/dirlibft
	#find ${PATH_LIBFT} -type f -name "*.[ch]" -print | xargs cp -t ${PATH_TEST}/dirlibft
	find ${PATH_LIBFT} -type f -name "*.[ch]" -exec cp {} ${PATH_TEST}/dirlibft  \;
	PATH_LIBFT=${PATH_TEST}/dirlibft
fi

for part in ${tab_all_part[*]}
do
	opt_part=$(echo OPT_NO_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
	if [ ${!opt_part} -eq 0 ]
	then
		p=0
		tab_part=$(echo ${part}[*])
		nb_func=$(echo ${!tab_part} | wc -w)
		if [ ${part} != "Additional_func" ]
		then
			while (( p < ${nb_func} ))
			do
				(( ${part}_activation[$p]=1 ))
				(( p += 1 ))
			done
		else
			(( ${opt_part}=1 ))
			while (( p < ${nb_func} ))
			do
				func_name=$(echo ${part}[$p])
				if [ -e ${PATH_LIBFT}/${!func_name} ]
				then
					(( ${part}_activation[$p]=1 ))
					(( ${opt_part}=0 ))
				fi
				(( p += 1 ))
			done
		fi
	fi
done

init_deepthought()
{
	if [ -e ${PATH_DEEPTHOUGHT}/deepthought ]
	then
		rm -f ${PATH_DEEPTHOUGHT}/deepthought
	fi
	text="= Host-specific information "
	printf "${text}" >> ${PATH_DEEPTHOUGHT}/deepthought
	printf "%.s=" $(seq 1 $(( 80 - ${#text} ))) >> ${PATH_DEEPTHOUGHT}/deepthought
	printf "\n$> hostname; uname -msr\n" >> ${PATH_DEEPTHOUGHT}/deepthought
	hostname >> ${PATH_DEEPTHOUGHT}/deepthought
	uname -msr >> ${PATH_DEEPTHOUGHT}/deepthought
	printf "$> date\n" >> ${PATH_DEEPTHOUGHT}/deepthought
	date >> ${PATH_DEEPTHOUGHT}/deepthought
	printf "$> gcc --version\n" >> ${PATH_DEEPTHOUGHT}/deepthought
	gcc --version >> ${PATH_DEEPTHOUGHT}/deepthought
	printf "$> clang --version\n" >> ${PATH_DEEPTHOUGHT}/deepthought
	clang --version >> ${PATH_DEEPTHOUGHT}/deepthought
}

clear
init_deepthought

if [ -e ${PATH_LIBFT}/Makefile ]
then
	MAKEFILE_VAR="Makefile"
elif [ -e ${PATH_LIBFT}/makefile ]
then
	MAKEFILE_VAR="makefile"
else
	MAKEFILE_VAR="missing_makefile"
fi

if [ ${OPT_NO_SEARCH} -eq 0 ]
then
	func_check_file
fi
if [ ${OPT_NO_LIBRARY} -eq 0 ]
then
	func_compil_lib
fi

for part in ${tab_all_part[*]}
do
	opt_part=$(echo OPT_NO_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
	activate_part=$(echo ACTIVATE_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
	if [ ${!opt_part} -eq 0 ]
	then
		(( ${activate_part}=1 ))
	fi
done

for part in ${tab_all_part[*]}
do
	activate_part=$(echo ACTIVATE_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
	if [ ${!activate_part} -eq 1 ]
	then
		text="DON'T WORRY, IF YOU GOT A SIGABRT ON THE LAST TEST OF MEMCHR THAT'S A NORMAL BEHAVIOR ! (LST's functions didn't work aswell sry)"
		printf "${COLOR_TITLE}"
		printf "%.s${CHAR_LENGTH}" $(seq 1 ${TITLE_LENGTH})
		printf "\n${CHAR_WIDTH}\033[$(( (${TITLE_LENGTH} - ${#text}) / 2 ))G${text}\033[${TITLE_LENGTH}G${CHAR_WIDTH}\n"
		printf "%.s${CHAR_LENGTH}" $(seq 1 ${TITLE_LENGTH})
		printf "\n${DEFAULT}"
		break
	fi
done

if [ -e ${PATH_LIBFT}/libft.h ]
then
	cp ${PATH_LIBFT}/libft.h ${PATH_TEST}
fi

printf "#include \"libft.h\"\n\nint\tmain(void)\n{\n\treturn (0);\n}" > ${PATH_TEST}/main_check_cheating.c

for part in ${tab_all_part[*]}
do
	activate_part=$(echo ACTIVATE_${part} | tr '[:lower:]' '[:upper:]' | rev | cut -c 6- | rev)
	if [ ${!activate_part} -eq 1 ]
	then
		text="= ${part}tions "
		printf "\n${text}" >> ${PATH_DEEPTHOUGHT}/deepthought
		printf "%.s=" $(seq 1 $(( 80 - ${#text} ))) >> ${PATH_DEEPTHOUGHT}/deepthought
		printf "\n" >> ${PATH_DEEPTHOUGHT}/deepthought
		test_function $(echo ${part}[*])
	fi
done

if [ -e ${PATH_TEST}/a.out ]
then
	rm ${PATH_TEST}/a.out
fi

if [ -e ${PATH_TEST}/libft.h ]
then
	rm ${PATH_TEST}/libft.h
fi

if [ -e ${PATH_TEST}/main_check_cheating.c ]
then
	rm ${PATH_TEST}/main_check_cheating.c
fi

if [ ${DIRECTORY} -eq 1 ]
then
	if [ -d ${PATH_TEST}/dirlibft ]
	then
		rm -rf ${PATH_TEST}/dirlibft
	fi
fi

if [ ${ACTIVATE_PART1} -eq 1 ] || [ ${ACTIVATE_PART2} -eq 1 ] || [ ${ACTIVATE_BONUS} -eq 1 ] || [ ${ACTIVATE_ADDITIONAL} -eq 1 ]
then
	printf "Abort : ${RED}A${DEFAULT} Bus error : ${RED}B${DEFAULT} Segmentation fault : ${RED}S${DEFAULT} Timeout : ${RED}T${DEFAULT} Nothing turned in : ${RED}NTI${DEFAULT}\n"
	printf "\n"
fi
printf "A deepthought file has been generated in ${COLOR_DEEPTHOUGHT_PATH}${PATH_DEEPTHOUGHT}\n\n${DEFAULT}"
make --no-print-directory -C ${PATH_LIBFT} fclean > /dev/null
